### PR TITLE
Add ResultType discriminator to RecognizeResult and ReasonCode casting

### PR DESCRIPTION
**Pull request by atazimsft**
This pull request:
- Creates a RecognizeResultType struct to be used as a discriminator for RecognizeResult and its subclasses, ChoiceResult and CollectTonesResult, in order to make the types more discoverable via intellisense.

- Adds a GetReasonCodeValue() function to the ReasonCode struct to allow users to easily access the integer subcode
For example:
`int reasonCodeValue = playFailedEvent.ReasonCode.GetReasonCodeValue(); // 8536`

### CODE REVIEW

**Comment by wangrui-msft** [comment]
Looks good to me, just need to run a api export on local to unblock the build pipeline.

**Review by wangrui-msft — APPROVED** [review]

**Review by cochi2 — COMMENTED** [review]
overall it looks good, I have a couple of comments:
* what is the use case to have `GetReasonCodeValue`? BTW, we could also get the same value from the actual `RecognizeFailed` event through `resultInformation.SubCode`, is the reason to have this new method to help discoverability of such variable?
* Tests are missing, at least unit tests should be here to test that this works.

File: sdk/communication/Azure.Communication.CallAutomation/src/Models/RecognizeResultType.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+using System;
+using System.ComponentModel;
+
+namespace Azure.Communication.CallAutomation.Models
+{
+    /// <summary>
+    /// Determines the sub-type of the recognize result.
+    /// </summary>
+    public readonly partial struct RecognizeResultType: IEquatable<RecognizeResultType>
```
Comment (arupdutta-msft) [inline sdk/communication/Azure.Communication.CallAutomation/src/Models/RecognizeResultType.cs:12]: Any reason that we cannot use enum here?

Comment (cochi2) [reply sdk/communication/Azure.Communication.CallAutomation/src/Models/RecognizeResultType.cs:12]: Do we already know all the possible values this RecognizeResultType could get? If there is a chance that in the future there will be more types, we have to use the struct.

**Review by cochi2 — APPROVED** [review]

**Pull request merged**